### PR TITLE
Support ResNet50-int8 model in ImageClassifier

### PR DESCRIPTION
**Pull request by hkvision**
- Run calcScales with `resnet-50.bigdl` to get `resnet-50.quantized.bigdl`
- Convert to ZooModel `analytics-zoo_resnet-50-int8_imagenet_0.5.0.model` and add corresponding preprocessor in ImageClassifier.
- ImageNetInference example for accuracy test.
- Perf for performance test. Throughput should be correct accoriding to @i8run 
- Predict example, which is the same as ImageClassification except here we use int8 model.

Here for BigDL int8 example: https://github.com/intel-analytics/BigDL/tree/branch-0.8/spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8

### CODE REVIEW

**Comment by hkvision** [comment]
cc @i8run

**Comment by hkvision** [comment]
@jason-dai Any comments?

File: zoo/src/main/scala/com/intel/analytics/zoo/models/image/common/ImageModel.scala [hunk 1]
```diff
@@ -85,9 +108,12 @@ object ImageModel {
...
    * @return
    */
-  def loadModel[T: ClassTag](path: String, weightPath: String = null, modelType: String = "")
+  def loadModel[T: ClassTag](path: String, weightPath: String = null, modelType: String = "",
+                             quantize: Boolean = false)
     (implicit ev: TensorNumeric[T]): ImageModel[T] = {
-    val model = Module.loadModule[T](path, weightPath)
+    var model = Module.loadModule[T](path, weightPath)
+    val modelName = model.getName()
+    if (quantize) model = model.quantize()
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/models/image/common/ImageModel.scala]: why do we need to quantize on the fly?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/image/ImageSetToSample.scala [hunk 2]
```diff
@@ -30,7 +33,7 @@ class ImageSetToSample[T: ClassTag](inputKeys: Array[String] = Array(ImageFeatur
                        targetKeys: Array[String] = null,
                        sampleKey: String = ImageFeature.sample)(implicit ev: TensorNumeric[T])
   extends ImageProcessing {
-  private val internalTransformer = ImageFrameToSample[T](inputKeys, targetKeys, sampleKey)
+  private val internalTransformer = InternalImageFrameToSample[T](inputKeys, targetKeys, sampleKey)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/image/ImageSetToSample.scala]: Why do you implement an `InternalImageFrameToSample`? Why not implement the new features directly in `ImageSetToSample`?

Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/image/ImageSetToSample.scala]: This applies to all the new ImageProcessing above.

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/image/ImageSetToSample.scala]: Ar, I added this just in order to quickly work out the solution. I will refactor this.
The purpose of this change is: Since each ImageModel uses the same preprocessor, for prediction we don't have label and for evaluation we have label. Update this transformer to support both cases.

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/image/ImageSetToSample.scala]: Updated. The change wouldn't affect existing ImageModels.

File: zoo/src/main/scala/com/intel/analytics/zoo/models/image/common/ImageModel.scala (hunk 1 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/models/image/common/ImageModel.scala]: Removed quantize. Detect whether the model has weight scales based on SpatialConvolution layers. Please take a look :)

**Comment by hkvision** [comment]
cc @jason-dai This PR has been finished.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/ImageNetInference.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        .text("The path to the int8 quantized ResNet50 model snapshot")
+        .action((x, c) => c.copy(model = x))
+        .required()
+      opt[Int]('b', "batchSize")
+        .text("batch size")
+        .action((x, c) => c.copy(batchSize = x))
+    }
+    parser.parse(args, ImageNetInferenceParams()).map(param => {
+      val sc = NNContext.initNNContext("ImageNet2012 with Int8 Inference Example")
+      val images = ImageSet.readSeqFiles(param.folder, sc)
```
Comment (qiyuangong) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/ImageNetInference.scala]: Pls add more details or comments about this Sequence Files (image format), such that user can reproduce our results in their environment.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Perf.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      val model = ImageClassifier.loadModel[Float](param.model)
+      model.setEvaluateStatus()
+
+      var iteration = 0
+      while (iteration < param.iteration) {
+        val start = System.nanoTime()
+        model.forward(batchInput)
+        val timeUsed = System.nanoTime() - start
+        val throughput = "%.2f".format(batchSize.toFloat / (timeUsed / 1e9))
+        logger.info(s"Iteration $iteration, takes $timeUsed ns, throughput is $throughput imgs/sec")
```
Comment (qiyuangong) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Perf.scala]: Pls add more descriptions about batch and single record.

File: zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala [hunk 5]
```diff
@@ -129,6 +133,13 @@ object ImagenetConfig {
     commonPreprocessor(Consts.IMAGENET_RESIZE, 224, 123, 117, 104)
   }
 
+  def resnetNewPreprocessor(): Preprocessing[ImageFeature, ImageFeature] = {
```
Comment (qiyuangong) [inline zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala]: I think `resnetNewPreprocessor` is not a good method name. :)

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala]: changed to `bigdlResNetPreprocessor`

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/ImageNetInference.scala (hunk 3 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/ImageNetInference.scala]: Added code doc. Will add more in README.

File: zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala [hunk 6]
```diff
@@ -125,10 +129,19 @@ object ImagenetConfig {
...
+  // Preprocessor for ResNet50 pre-trained in Caffe
   def resnetPreprocessor() : Preprocessing[ImageFeature, ImageFeature] = {
     commonPreprocessor(Consts.IMAGENET_RESIZE, 224, 123, 117, 104)
   }
 
+  // Preprocessor for ResNet50 pre-trained in BigDL
+  def bigdlResNetPreprocessor(): Preprocessing[ImageFeature, ImageFeature] = {
+      ImageRandomResize(256, 256) ->
+      ImageRandomCropper(224, 224, mirror = false, cropperMethod = CropCenter) ->
+      ImageChannelScaledNormalizer(104, 117, 123, 0.0078125) ->
```
Comment (qiyuangong) [inline zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala]: In normal image net precessing, RGB means are `[123, 117, 104]`. But, in this bigDLResnet. It is reversed. Does it means the input image data are in BGR format?

**Comment by qiyuangong** [comment]
LGTM. :)

**Comment by hkvision** [comment]
Merge it first.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Perf.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        .text("The number of iterations to run the perf test. " +
+          "The result will be the average of each iteration time cost")
+        .action((v, p) => p.copy(iteration = v))
+    }
+
+    parser.parse(args, PerfParams()).foreach { param =>
+      val batchSize = param.batchSize
+      val batchInput = Tensor(Array(batchSize, 3, 224, 224)).rand()
+      val singleInput = Tensor(Array(1, 3, 224, 224)).rand()
+      Engine.init
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Perf.scala]: This examples runs in local mode and doesn't actually need `sc` but only needs to init the Engine. I don't use the command `spark-submit` but just uses `java -cp ...`

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Perf.scala [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        val start = System.nanoTime()
+        model.forward(batchInput)
+        val timeUsed = System.nanoTime() - start
+        val throughput = "%.2f".format(batchSize.toFloat / (timeUsed / 1e9))
+        logger.info(s"Iteration $iteration, batch $batchSize, takes $timeUsed ns, " +
+          s"throughput is $throughput imgs/sec")
+        iteration += 1
+      }
+
+      val model2 = ImageClassifier.loadModel[Float](param.model)
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Perf.scala]: Seems not. For an int8 model in BigDL, for the first forward if the batch size is fixed, then in later forward it will always use this batch. Thus the first model calcs throughout for a specific batch while we need another model to calc latency for batch=1.

File: zoo/src/main/scala/com/intel/analytics/zoo/models/image/common/ImageModel.scala [hunk 9]
```diff
@@ -67,6 +70,36 @@ abstract class ImageModel[T: ClassTag]()(implicit ev: TensorNumeric[T])
...
+      vMethods: Array[_ <:ValidationMethod[T]],
+      batchSize: Int,
+      configure: ImageConfigure[T] = null): Array[(ValidationResult, ValidationMethod[T])] = {
+    val evalConfig = if (null == configure) config else configure
+    // Remark: BigDL currently only supports evaluation on DistributedImageSet.
+    if (evalConfig == null) {
+      model.evaluateImage(image.toImageFrame(), vMethods, Some(batchSize))
+    } else {
+      // Remark: ImageConfigure only has batchPerPartition while evaluate needs
+      // total batchSize. Thus add a batchSize argument here first.
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/models/image/common/ImageModel.scala]: Sure. If config is null, batchPerPartition defaults to be 4?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Predict.scala [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+      val result = labelOutput(output).toLocal().array
+
+      logger.info(s"Prediction result")
+      result.foreach(imageFeature => {
+        logger.info(s"image : ${imageFeature.uri}, top ${param.topN}")
+        val clses = imageFeature("clses").asInstanceOf[Array[String]]
+        val probs = imageFeature("probs").asInstanceOf[Array[Float]]
+        for (i <- 0 until param.topN) {
+          logger.info(s"\t class: ${clses(i)}, credit: ${probs(i)}")
+        }
```
[the remark this answers is not in the document]
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/bigdl/Predict.scala]: Yeah. This example refers to original image classification predict. While the for the new BigDL ResNet50 model, it needs to add a SoftMax to generate probabilities. And also, the vnni Predict example is a local example mainly for observing performance boost. Any suggestions how to unify the two examples (not that straightforward to me lol)...

File: zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala (hunk 6 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/models/image/imageclassification/ImageClassificationConfig.scala]: cc @dding3
